### ISSUE TITLE
Donating items/Donating towards goals

### ISSUE THREAD

**Issue by pjeweb** [alveusgg/alveusgg#328]
It would be great to have a way for users to donate items or donate towards a goal set by Alveus.

Ideas:
- Donate items not buy-able or shipable through the Amazon Wishlist => Users donate the sum, but do not buy or ship the item. Instead alveus buys it themselves. This way items that cant be shipped to the PO Box can be donated. The site would track the user and allows alveus to attribute the donated item to them and shout them out on the PO Box stream.
- Donate expensive items collectively: To make bigger donations like the cow brush possible, allow users to donate partial sums attributed to that item. This would probably need some minimums (total sum and partial donation) such that only a handful of people get a shout out on the stream.
- Donate towards specific goals (projects/enclosures) or for specific ambassadors. Right now there are no plans to reward those specificly, but it might be possible to send them digital or physical gifts for their donation (maybe specific for the ambassador)

Tech ideas:
- Set up webhooks for stream overlays
- Implement a backend for alveus staff to manage goals/items
- Pay via PayPal? Maybe Paypal Giving Fund so users get a receipe?
- Send notifications/emails when items/goals are reached to participants and alveus staff.
- Needs a simple way for alveus to see attributions for PO Box streams.

Problems:
- check regulations for donations / non-profits
- bookkeeping requirements

**Comment by nitelite** [comment]
Donation goals are fairly common in streaming. To drive viewer interaction on the site, you could have a separate OBS scene that can be triggered by mods/staff which swap out the live cams entirely to run a short "ad" for the current donation goals. 

The scene could pull in an image, description, possibly a short video recorded by staff when the item is added, to promote the donation drive. Could also include the top donators for that item, combined with the link to the donation page being posted in chat if someone wants to join in.

I think the same "short, partially-automated ad in-stream" concept might work well for promoting upcoming events, so maybe if its built in a decently generic way, it could also be used to promote other stuff (like Summer Camp) with a countdown.

**Comment by pjeweb** [comment]
> I think the same "short, partially-automated ad in-stream" …

Yeah, great idea we had on our radar. I wrote down what we had and used added you suggestions! I especially like the idea with the dono goals and event countdown! #329

**Comment by MattIPv4** [comment]
I think the primary goal here, initially, should be to get the site accepting donations directly via PayPal and receiving notifications for those payments via IPN. I agree it'd be best to use the giving feature if we can, as I believe there are less fees, and the user gets a proper donation receipt.

Once we have donations working with payment notifications, tying those notifications back to a specific goal in the DB etc. should be relatively simple.

**Comment by Masternater** [comment]
Sorry, I know I'm the "new guy" jumping in...

I think another thing that could work well with the "progress bar" type donations, would be funding more mundane day-to-day items.  I think people would really like to be able to see that they are partially funding a load of hay for the pasture, or buying a bag of feed for a particular ambassador.  

Obviously, a lot of that stuff gets bought regardless...so maybe the progress bar could show the "general" Alveus fund gradually being offset by the viewers immediate contribution.  And, I realize the costs aren't fixed...so some number fudging might be required....previous month bill, average costs, or just a target safely below the cost (on the assumption the general fund makes up the difference).

**Comment by nitelite** [comment]
@MattIPv4 and @pjeweb 
What are you thoughts around buiding out services like this? Would functionality like the IPN receiver be built as part of the site repo or would you prefer to have more of a "service architecture" where the IPN receiver has its own repo, hosted as a separate service, providing an API for other parts, like the site etc?

**Comment by MattIPv4** [comment]
I think for consistency, it'd be part of the existing API: https://github.com/alveusgg/alveusgg/tree/main/apps/website/src/pages/api

I generally prefer having services split out from the website, and dislike that we have an API buried within the "pages" of the frontend site, but given that's the pattern we have, we should probably stick to it for consistency.

**Comment by pjeweb** [comment]
I do not really have strong personal feelings about that. But I think I would not want to have to operate a complex service architecture in addition to what we have.

I also do not really mind having it in one repo, nor do i mind next. If we stick with next, we can take a look at route handlers in the "new" app directory. Then it would at least not be under pages ;) https://nextjs.org/docs/app/building-your-application/routing/router-handlers

I do not really have any Ops experience beyond this here and old-school LAMP on root servers (maybe with some CDN or simple load balancing). This is kinda getting off-topic, but i would like to check out if we can maybe set up something besides Vercel to host "everything". Maybe serverless with SST.dev/OpenNext on AWS so we do not have that many service providers for DB=Planetscale, S3=DO, Queue=Upstash, Compute+CDN=Vercel. I would personally not be comfortable setting up k8s or cdk/terraform or something of that caliber without heavy help.

**Comment by nitelite** [comment]
I started looking into this yesterday, getting everything set up for development and started looking into the PayPal part of it. 

Based on what I am finding, it seems like they are moving away from the "classic" IPN mechanism for notifications and trying to move developers over to their new "webhooks" mechanism instead. While it doesn't say IPN are becoming deprecated (and to be fair, I don't think they would deprecate it, see how many people probably depend on it), but they are certainly moving in that direction. 

For instance, the IPN simulator on their developer pages is apparently no longer being supported after December 8th. Based on the discussions people are having online, it also seems like webhooks "react" to payments faster than IPN notification do.

![image](https://github.com/alveusgg/alveusgg/assets/1143814/6d119297-721b-44c7-a5bb-bcb4093e2de4)

Unless you have any objections, I think I will focus on the webhooks mechanism as that seems to be PayPal's preferred approach these days.

**Comment by MattIPv4** [comment]
My knowledge of PayPal is super outdated, so if something (webhooks) has replaced IPN, sounds fine to me to use that instead 👍

**Comment by nitelite** [comment]
Ok, so a little update. After getting the site up and running locally, making a simple webhook receiver and verifying that it can parse incoming hooks using the webhook simulator with ngrok, I started looking into testing it out with the "full-blown" PayPal sandbox account system. 

They basically have all of www.paypal.com available at www.sandbox.paypal.com, which is a test-version of PayPal. This should allow me to create a sandbox business, create a donation link for that sandbox business and then donate "sandbox money" from a sandbox personal account to the sandbox business through the link.

After fiddling around with this for most of yesterday, it looks like webhooks are not yet supported through the donation flow. I am at least not able to get the sandbox to push me any webhook when I use those. There seems to be other people with similar issues, ex: ttps://www.reddit.com/r/webdev/comments/wcwe06/how_to_receive_webhooks_from_paypal_donations_and/

After this apparent set back, I also made a quick IPN implementation and that seems to be working fine for donation links/buttons. One challenge with IPN seems to be that it's only possible to define one IPN URL for one business acccount. I assume that IPN might already be in use by the overlays that show donation or something like that? 

This adds another layer of complexity. In theory we could either use a service like Zapier or build our own "IPN proxy", which receives one IPN message and relays it out to multiple defined address, but this feels like a bit of a hack? Anyone got any suggestions/insights/ideas?

**Comment by MattIPv4** [comment]
https://developer.paypal.com/api/nvp-soap/ipn/IPNSetup/#link-dynamicallysettingthenotificationurl would imply you can set an IPN url per button?

**Comment by MattIPv4** [comment]
Have you looked at PayPal's charity flow at all? I wonder if we could use that instead of the standard payment flow, so that folks get a proper charity receipt? Is that compatible with IPN/webhooks?

**Comment by MattIPv4** [comment]
> I have looked at what they call "donation buttons/links" mainly, as I assumed like you mentioned, that the receipts generated by this flow might be different from the typical checkout flow. Do you think that is the same as the one you are thinking of or is there a separate charity flow I might not be familiar with?

I'm thinking of their Giving Fund, which generates actual donation receipts (which is what Twitch uses for their new charity feature). I believe donation buttons/links don't generate anything special, beyond a regular PayPal payment.

> As a possible solition: we could overwrite the IPN URL using the notify_url parameter for the donation buttons on the site, add the item_number corresponding to the donation goal id the user chose before starting the donation and then, after we process that IPN notification, we forward the exact IPN request to the configured IPN URL, so that it also appear on stream as well?

Yah, if IPN is already being used for the stream overlay (I don't think it is -- I believe SE handles it all within their own PayPal flow, as https://support.streamelements.com/en/articles/166-paypal-tipping-system + https://support.streamelements.com/en/articles/38-tipping-system-overview-including-se-pay give no indication of needing to configure anything in PayPal), then we could set up forwarding.

**Comment by nitelite** [comment]
Lots of useful info, thanks 👍 Donation links seem to very similar flow to normal payment, only with slightly different wording (donate vs pay) in the flow, yeah.

I will look into Giving Fund and IPN then, I already have the IPN receiver ready and working for standard donation links, including passing info about what "donation goal" you would like your donation to go towards. Will check if the Giving Fund has similar functionality.

**Comment by nitelite** [comment]
I have been scratching my noggin the last few days trying to figure out how to create a charity account in the PayPal sandbox. After posting on the PayPal developer support forums, they responded with instructions that helped me realize that non-profit business accounts are only supported on the sandbox when the sandbox business accounts is created with country set to "USA".

Hopefully I should be able to set up a proof of concept in the PayPal sandbox in the next few days.

**Comment by MattIPv4** [comment]
Oh that's fun, what a silly arbitrary limitation. Appreciate you digging into this so thoroughly! 💚

**Comment by nitelite** [comment]
A bit of an update. Was away for a few weeks in Tokyo, but back in business now. Created a forum post on the paypal forum to see if I can get my sandbox account verified as a PayPal Giving Fund charity and test the IPN code with the donation system.

For reference later: https://www.paypal-community.com/t5/Sandbox-Environment/Enroll-test-charity-into-PayPal-Giving-Fund-program-for-testing/m-p/3090555#M11136

**Comment by nitelite** [comment]
No dice with the support forum. I have created a support ticket to see if they can get my test business added to the giving fund list, so I can actually test that it all works.

**Comment by nitelite** [comment]
@pjeweb 
I have kinda given up on getting Giving Fund transactions working in their sandbox.

Do you think it would be an option to get someone to set up IPN on the actual Alveus paypal account, point it at an endpoint on the website that logs data about the payload and do a Giving Fund transaction? I need to see what data PayPal provides for one of those transactions and looks like its hard to get anyone at PayPal to approve my test charity in their sandbox.

**Comment by pjeweb** [comment]
> Do you think it would be an option to get someone to set up IPN on the actual Alveus paypal account, point it at an endpoint on the website that logs data about the payload and do a Giving Fund transaction?

Yeah I asked Space. As said in chat it might be the easiest if we can give him a nodejs script he can put on a server they control so he does not have to share the secrets etc. How hard do you think would it be to use a distributed architecture like this for production as well? I imagine logging the IPNs is a pretty light task we could run on a single machine by alveus.  That could call an API provided by the site than for processing it further?! Does that make sense or am I misunderstanding something?

Any good pointers what to look at in terms of documentation for setting up the paypal account on their side? (And maybe for me to understand the IPN process in general?)

**Comment by pjeweb** [comment]
Quick note: Space mentioned that Giving Fund might not actually be preferred as it seems to not be expected by paypal for non-profits to use the giving fund to directly donate to them. It's rather a platform to donate, while "recommending" a charitable org the money *should* go to. They also retain the money for months before paying it out. That maybe bad for the donation goals. So maybe wait before doing any more work on the giving fund thingy. If I hear back from Space if we want to "just" use the normal paypal process instead or not, I'll let you know.

**Comment by pjeweb** [comment]
Prior art: https://throne.com/

**Comment by Masternater** [comment]
Pretty interfaces are nice, but what if it was set up like the feeder...where someone just has to include a "!" phrase in their donation message?  And then have the bot occasionally post the goals and current progress to chat...or prompted like the other commands?  Sure, the money would be going into the same "pool", but it'd just be a bit of an "honor system" that staff would be using the money as the donors intended.  And...maybe with a little more urgency than demonstrated with the Subathon goals (I know, I know...extenuating circumstances...but still not a great look)

It's a little more of a down-and-dirty approach...but it'd probably be good enough to accomplish what it needs to.  It'd also be a good test to see if it's a system that is worth devoting more time to.

**Comment by pjeweb** [comment]
> Pretty interfaces are nice, but what if it was set up like the feeder...where someone just has to include a "!" phrase in their donation message? And then have the bot occasionally post the goals and current progress to chat...or prompted like the other commands? Sure, the money would be going into the same "pool", but it'd just be a bit of an "honor system" that staff would be using the money as the donors intended. And...maybe with a little more urgency than demonstrated with the Subathon goals (I know, I know...extenuating circumstances...but still not a great look)
> 
> It's a little more of a down-and-dirty approach...but it'd probably be good enough to accomplish what it needs to. It'd also be a good test to see if it's a system that is worth devoting more time to.

That's a valid approach too. The question would be who/what is tracking the goals and who manages the !goals. Ideally anyone one the alveus team can add goals and they dont have to track it manually by going through donations. So we probably need to build something anyways. It's not that much more complicated what we want to do. Just another way to input the goals (besides stream elements) and automatically tracking the !goals when donating instead of doing it afterwards through the paypal receipts (or even manually). I hope that makes sense.

I do not really want to recreate throne btw. Just something i stumbled over that we might be able to get some inspiration from. I do like that the shops pay for the platform and they take care of shipping tho, without donators having to handle it. So its not uselesss.

**Comment by Masternater** [comment]
Oh, I wasn't dismissing the value of looking at other examples.  I'm just trying to think of ways to get something, _anything_,  up and running quickly.  Operating on the assumption that there's already software structure in place to both recognize "!feed", and the amount of that donation...it seemed to me that could be expanded on. Eg, !item1 <amount>, !item2 <amount>, !item3 <amount>...with a little bit of automated accounting, transferred to a bot notification in chat.

Honestly...I do not see where there would be an onerous amount of management overhead.  It's just changing a few item and goal values, once a month or so.  If they wanted to change the !feed amount to $10, right now...I can't imagine that wouldn't be trivial.

Something something...perfection getting in the way of "good enough".

**Comment by umami-umami** [comment]
> I do not really want to recreate throne btw. Just something i stumbled over that we might be able to get some inspiration from. I do like that the shops pay for the platform and they take care of shipping tho, without donators having to handle it. So its not uselesss.

Hey! Just curious, why not use Throne for this feature?

**Comment by Moolash3** [comment]
Id be interested in giving this a shot if this is still wanted/prio. Id highly recommend only using paypal and or stripe and avoid giving fund as its pretty frustrating to work with and isnt a "direct" donation.

I have a few questions/edge cases

- How do refunds work both before completion and after? My assumption would be that total item raised value would go down if it hasnt met the goal and if it has but has not been purchased yet it could get relisted or the cost just is eaten.
- Can an item be delisted after it has progress?
- If someone start the payment process toward an item but the item gets completed before the users payment clears, is that just considered a donation at that point with no item attribution? In the future could maybe give them some sort of credit to attribute their donation with another item but might be to much for an MVP
- Any interest in allowing users to donate publicly? or would it always just be x donators donated to this item, type of thing?

**Comment by MattIPv4** [comment]
@pjeweb @changesbyjames I think there were some recent internal discussions around how we want to approach this. Is that context able to be shared and is this something we can get the community to help with building?
